borsh/src/schema_helpers.rs

```diff
@@ -36,3 +36,208 @@ pub fn schema_container_of<T: BorshSchema>() -> BorshSchemaContainer {
 
     BorshSchemaContainer::new(T::declaration(), definitions)
 }
+
+/// Returns the largest possible size of serialised object.
+///
+/// Returns `None` if the maximal size overflows `usize`.  This may happen for
+/// nested dynamically-sized types such as `Vec<Vec<u8>>` whose maximum size is
+/// `4 + u32::MAX * (4 + u32::MAX)`.
+///
+/// The function has limitations which may lead it to overestimate the size.
+/// For example, hypothetical `IPv4Packet` would be encoded as at most ~64 KiB.
+/// However, if it uses sequence schema, this function will claim that the
+/// maximum size is ~4 GiB.
+///
+/// # Example
+///
+/// ```
+/// assert_eq!(Some(0), borsh::max_serialized_size::<()>());
+/// assert_eq!(Some(8), borsh::max_serialized_size::<usize>());
+///
+/// // 4 bytes of length and u32::MAX for the longest possible string.
+/// assert_eq!(Some(4 + 4294967295), borsh::max_serialized_size::<String>());
+///
+/// assert_eq!(None, borsh::max_serialized_size::<Vec<String>>());
+/// ```
+pub fn max_serialized_size<T: BorshSchema>() -> Option<usize> {
```

Comment:
Instead of `Option<usize>` it might return a newly declared enum:

```rust
pub enum Size {
    Bytes(usize),
    OverflowsUsize,
    /// a type being recursive means that the max size of a type is potentially infinite in some pure sense,
    /// and probably too hard to compute in practice if computed based on rust's thread stack size as input parameter
    Recursive,
}
```